### ISSUE TITLE
Sized versions of duk_{get,require}_[u]int()

### ISSUE THREAD

**Issue by fatcerberus** [svaarala/duktape#1059]
API documentation for `duk_get_uint()` says:

> The coercion is also different from Ecmascript ToUint32() coercion because the full range of the native duk_uint_t is allowed (which may be more than 32 bits). 

This is convenient in most cases, but there were times I wanted to ensure the value is clamped to exactly 32 bits as if by ToUint32(), regardless of the native int size (which is 64 bits on some platforms).  So guaranteed-size versions of these APIs would be nice to have.

**Comment by svaarala** [comment]
Hmm, `ToUint32()` doesn't clamp it's result but rather applies a modulo operation.

That's why I think `duk_get_uint()` etc should be renamed to `duk_get_uint_clamped()` or something because that clamping behavior differs from the usual integer coercions in Ecmascript so it's a bit surprising.

But taking this into account, you'd like to have e.g. `duk_get_uint32_clamped()`?

**Comment by fatcerberus** [comment]
Yeah, that's what I meant.  I should have realized ToUintX() operations were modulo since they are used with typed arrays :)

**Comment by svaarala** [comment]
Typed arrays, and in ES5, bitwise operations.

**Comment by fatcerberus** [comment]
Regarding the clamping behavior: Despite Ecmascript semantics, I think clamping makes the most sense for the public API, so I don't know that the API calls need a `_clamped` suffix in their names.  I can't vouch for others, but I can say that the current behavior wasn't surprising to me at least.  I would have found modulo behavior to be more surprising, to be honest.

**Comment by svaarala** [comment]
I'm sure expectations differ here. I don't think clamping is odd behavior as such, but I find it inconsistent that `duk_to_uint32()` is modulo based and `duk_to_uint()` is clamping. From outward appearances one might conclude that the only difference is the range or final type only which isn't the case.

**Comment by fatcerberus** [comment]
I guess the reason I find the modulo behavior odd is that I generally think of typical Ecmascript semantics where overflow = infinity.  The wrapping semantics are used only in those scenarios where that makes more sense (typed arrays and bitwise ops as you point out), otherwise saturation is the "default" behavior.

**Comment by fatcerberus** [comment]
That said, I don't really mind if you want to rename the API calls.  I'd just like sized versions. :)

**Comment by fatcerberus** [comment]
Here's one use case in my `system.sleep()` API implementation:
https://github.com/fatcerberus/minisphere/blob/master/src/engine/pegasus.c#L1216-L1220

That code will break if the user passes a value greater than 4G (admittedly probably not the dominant use case :) and the native int size is 64-bit.  Here clamping to 32-bit is preferred, but to do that I currently have to:
```c
double value = duk_require_number(ctx, 0);
uint32_t timeout = (uint32_t)(value < 0xffffffff ? value : 0xffffffff);
```

Which is more code, an extra throwaway variable, and includes an awkward-looking double->int cast.

**Comment by svaarala** [comment]
Yeah :) I'm not yet sure how to proceed so need to ponder a bit.

One approach would to add variants for `duk_to_uint32_modulo()` and `duk_to_uint32_clamped()` so that it would be explicit in the call site. Then just map the existing call using a macro and maybe deprecate it.

But then... if the API gets an explicit clamped API call, it'd also be natural to allow the clamp limits to be defined: they're not always the integer limits. For example, working with a timeout value it would be convenient to say:

```c
long timeout = (long) duk_to_int_clamped(ctx, 1, 10, 20000); /* Clamp timeout to [10ms, 20000ms] */
```

**Comment by svaarala** [comment]
There's also a small semantic difference between `duk_to_uint32()` and `duk_to_uint()` by the way: the latter does a `duk_to_number()` coercion followed by clamping and integer conversion. However, the value on the value stack is the `duk_to_number()` coerced value, i.e. may contain fractions, be NaN, etc.

`duk_to_uint32()` is Ecmascript defined and the coercion result is also stored on the value stack.

One could try to make `duk_to_uint()` to reutrn the same value as is left on the value stack. But for 64-bit values `DUK_INT_MAX` actually cannot be represented as an IEEE double, so that couldn't be then used as a clamp limit! One could return it from the API but a different value (close to it) would be left on the value stack. This would also be very confusing :)
